test/framework/test_options.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2020 Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package framework
+
+import (
+	"context"
+	"github.com/onsi/ginkgo"
+	"time"
+)
+
+// Test options contains options to add additional functionality
+// or cleanup handlers to a testcase.
+type TestOptions struct {
+	// AfterTests holds a list of all registered AfterTest functions
+	// that are executed when the test has finished.
+	AfterTests []func()
+
+	// AfterTests holds a list of all registered contextified AfterTest functions
+	// that are executed when the test has finished.
+	CAfterTests []CAfterTestOptions
+}
+
+// ApplyOptions applies the given test options on these options.
+func (o *TestOptions) ApplyOptions(opts []TestOption) *TestOptions {
+	for _, opt := range opts {
+		opt.ApplyToTest(o)
+	}
+	return o
+}
+
+// Complete registers all test options that are configured.
+// it should be a function that configures a ginkgo test case
+// This should get called when all options are applied.
+func (o *TestOptions) Complete(it func()) {
+
+	if len(o.AfterTests) == 0 && len(o.CAfterTests) == 0 {
+		it()
+		return
+	}
+
+	// Create a new context so that the AfterTests function only runs once after this one testcase.
+	// Otherwise the after tests would run after every test case in the outer context.
+	ginkgo.Context("", func() {
```

Comment:
Hmmm, this will add a "" line to ginkgo failure message, right? If we can't avoid it (which I assume), can we at least give it a name (like "AfterTest wrapper context" or whatever) to not add more confusion to failure messages?

Reply:
The context text does not really appear in the output.
It would just add the text after the text of `Describe`

See this exmaple:
```go
func TestFramework(t *testing.T) {
	RegisterFailHandler(Fail)
	RunSpecs(t, "Framework Test Suite")
}

var _ = Describe("test", func() {

	It("Without context", func(){
		Expect(true).To(BeTrue())
	})

	Context("my context", func() {
		It("With named context", func(){
			Expect(true).To(BeTrue())
		})
	})

	Context("", func() {
		It("With empty context", func(){
			Expect(true).To(BeTrue())
		})
	})

})
```

Would produce:
```
=== RUN   TestFramework
Running Suite: Framework Test Suite
===================================
Random Seed: 1588749141
Will run 3 of 3 specs

test 
  Without context
  /Users/d064999/go/src/github.com/gardener/test-infra/tmp/main_test.go:16
[It] Without context
  /Users/d064999/go/src/github.com/gardener/test-infra/tmp/main_test.go:16
•
------------------------------
test my context 
  With named context
  /Users/d064999/go/src/github.com/gardener/test-infra/tmp/main_test.go:21
[It] With named context
  /Users/d064999/go/src/github.com/gardener/test-infra/tmp/main_test.go:21
•
------------------------------
test  
  With empty context
  /Users/d064999/go/src/github.com/gardener/test-infra/tmp/main_test.go:27
[It] With empty context
  /Users/d064999/go/src/github.com/gardener/test-infra/tmp/main_test.go:27
•
Ran 3 of 3 Specs in 0.000 seconds
SUCCESS! -- 3 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestFramework (0.00s)
```

Reply:
Thanks for testing this!
As I am seeing the output from above now, I would probably leave the context string empty, as you have done it. It might be even more confusing to add a text there...